### PR TITLE
loader: make loader completely generic

### PR DESCRIPTION
**Pull request by rdmsr**

### CODE REVIEW

File: sources/embed/boot.h [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#pragma once
```
Comment (sleepy-monax) [inline sources/embed/boot.h:1]: This shouldn't be in embed/

File: sources/embed/mem.h [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include <bal/boot/handover.h>
```
Comment (sleepy-monax) [inline sources/embed/mem.h:3]: Don't include bal stuff in embed/

File: sources/libs/bal/hw/mem.h [hunk 3]
```diff
@@ -52,4 +52,3 @@ MaybeError bal_mem_paddr(BalMem *self, uintptr_t *paddr);
 void bal_mem_pack(IpcPack *pack, BalMem const *self);
 
 void bal_mem_unpack(IpcUnpack *pack, BalMem *self);
```
Comment (sleepy-monax) [inline sources/libs/bal/hw/mem.h:54]: Please follow our editor config.

File: sources/libs/brutal/parse/lex.h [hunk 4]
```diff
@@ -21,13 +21,14 @@ typedef struct
...
-        }                                                                                                        \
-    })
+#define lex_assert_no_error(LEX)                                                                                     \
+    (                                                                                                                \
+        {                                                                                                            \
+            if ((LEX)->has_error)                                                                                    \
+            {                                                                                                        \
+                panic$("lex contains error: '{}' with lexeme: '{}'", (LEX)->error.message, (LEX)->error.lexeme.str); \
+            }                                                                                                        \
+        })
```
Comment (sleepy-monax) [inline sources/libs/brutal/parse/lex.h:31]: Why

File: sources/libs/brutal/parse/scan.h [hunk 5]
```diff
@@ -31,14 +31,15 @@ typedef struct
...
-    })
+#define scan_assert_no_error(SELF)                                  \
+    (                                                               \
+        {                                                           \
+            if ((SELF)->has_error)                                  \
+            {                                                       \
+                scan_dump_error(SELF, io_chan_out());               \
+                panic$("scan: terminating due to previous errors"); \
+            }                                                       \
+        })
```
Comment (sleepy-monax) [inline sources/libs/brutal/parse/scan.h:42]: Why

File: sources/libs/brutal/ui/view.h [hunk 6]
```diff
@@ -40,11 +40,12 @@ struct _UiView
...
-    UiView *__view = (UiView *)alloc_make(alloc_global(), T); \
-    ui_view_init(__view);                                     \
-    __view;                                                   \
-})
+#define ui_view_create$(T) (                                      \
+    {                                                             \
+        UiView *__view = (UiView *)alloc_make(alloc_global(), T); \
+        ui_view_init(__view);                                     \
+        __view;                                                   \
+    })
```
Comment (sleepy-monax) [inline sources/libs/brutal/ui/view.h:48]: Why

File: sources/libs/stdc/math/libm.h [hunk 7]
```diff
@@ -205,11 +205,12 @@ union ldshape
...
-    ((union {               \
-         _Complex t __z;    \
-         t __xy[2];         \
-     }){.__xy = {(x), (y)}} \
+#define __CMPLX(x, y, t)     \
+    ((union                  \
+      {                      \
+          _Complex t __z;    \
+          t __xy[2];         \
+      }){.__xy = {(x), (y)}} \
```
Comment (sleepy-monax) [inline sources/libs/stdc/math/libm.h:213]: This is way out of scope of this PR

File: sources/loader/loader.h [hunk 8]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include <bal/boot.h>
```
Comment (sleepy-monax) [inline sources/loader/loader.h:2]: missing new line after `#pragma`

File: sources/loader/main.c [hunk 9]
```diff
@@ -1,21 +1,19 @@
...
-#include <efi/srvs.h>
-#include <efi/tty.h>
 #include <elf/elf.h>
-#include "loader/config.h"
-#include "loader/memory.h"
-#include "loader/protocol.h"
+#include <loader/config.h>
+#include <loader/loader.h>
+#include <loader/memory.h>
+#include <loader/protocol.h>
```
Comment (sleepy-monax) [inline sources/loader/main.c:8]: Only library includes should be wrapped in <>

File: sources/loader/efi/entry.c [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+    loader_entry();
+
+    panic$("loader_entry shouldn't return!");
+}
+
+void loader_boot_deinit(void)
+{
+    efi_deinit();
+}
```
Comment (sleepy-monax) [inline sources/loader/efi/entry.c:22]: useless function

**Review by sleepy-monax — CHANGES_REQUESTED** [review]

File: sources/loader/loader.h [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+// Handover functions, used by the loader to fill the boot protocol properly.
+void loader_handover_fill_mmap(HandoverMmap *self);
+
+void loader_handover_fill_framebuffer(HandoverFramebuffer *fb, uint64_t requested_width, uint64_t requested_height);
+
+uintptr_t loader_handover_get_rsdp();
+
+// Memory allocation
+enum loader_mem_flag
```
Comment (Supercip971) [inline sources/loader/loader.h:17]: Use a typedef:
```c
typedef enum
{

} LoaderMemFlag;
```

File: sources/loader/loader.h [hunk 12]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#include <bal/boot.h>
+
+// Entry of the loader, called by the platform's entry point. This is needed because of EFI requiring parameters to be passed to its entry point
```
Comment (Supercip971) [inline sources/loader/loader.h:5]: Maybe add doxygen doc instead of a comment

File: sources/loader/main.c [hunk 13]
```diff
@@ -101,33 +92,26 @@ void loader_boot(LoaderEntry const *entry)
...
     EntryPointFn entry_point = loader_load_kernel(entry->kernel, vmm);
 
-    log$("Kernel loaded, jumping in to it...");
     Handover *handover = allocate_handover(vmm);
-    efi_populate_handover(entry, handover);
 
-    efi_deinit();
+    loader_populate_handover(entry, handover);
+
     memory_switch(vmm);
```
Comment (Supercip971) [inline sources/loader/main.c:101]: The memory switch should be the last thing we do before jumping into the kernel.

**Review by Supercip971 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**